designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala

```diff
@@ -0,0 +1,147 @@
+package pl.touk.nussknacker.ui.process.newdeployment
+
+import cats.data.EitherT
+import db.util.DBIOActionInstances._
+import pl.touk.nussknacker.engine.api.deployment.DataFreshnessPolicy
+import pl.touk.nussknacker.engine.api.deployment.StateStatus.StatusName
+import pl.touk.nussknacker.engine.api.process.{ProcessIdWithName, ProcessName}
+import pl.touk.nussknacker.engine.deployment.ExternalDeploymentId
+import pl.touk.nussknacker.security.Permission
+import pl.touk.nussknacker.ui.db.entity.ProcessEntityData
+import pl.touk.nussknacker.ui.process.deployment.{
+  CommonCommandData,
+  DeploymentService => LegacyDeploymentService,
+  RunDeploymentCommand => LegacyRunDeploymentCommand
+}
+import pl.touk.nussknacker.ui.process.newdeployment.DeploymentEntityFactory.DeploymentEntityData
+import pl.touk.nussknacker.ui.process.newdeployment.DeploymentService._
+import pl.touk.nussknacker.ui.process.repository.{CommentValidationError, DBIOActionRunner, ScenarioMetadataRepository}
+import pl.touk.nussknacker.ui.security.api.LoggedUser
+
+import java.sql.Timestamp
+import java.time.Clock
+import scala.concurrent.{ExecutionContext, Future}
+
+// TODO: This class is a new version of deployment.DeploymentService. The problem with the old one is that
+//       it joins multiple responsibilities like activity log (currently called "actions") and deployments management.
+//       Also, because of the fact that periodic mechanism is build as a plug-in (DeploymentManager), some deployment related
+//       operations (run now operation) is modeled as a CustomAction. Eventually, we should:
+//       - Split activity log and deployments management
+//       - Move periodic mechanism into to the designer's core
+//       - Remove CustomAction
+//       After we do this, we can remove legacy classes and fully switch to the new once.
+class DeploymentService(
+    scenarioMetadataRepository: ScenarioMetadataRepository,
+    deploymentRepository: DeploymentRepository,
+    // TODO: we shouldn't call legacy service, instead we should new service from the legacy one
+    legacyDeploymentService: LegacyDeploymentService,
+    dbioRunner: DBIOActionRunner,
+    clock: Clock
+)(implicit ec: ExecutionContext) {
+
+  def processCommand(
+      command: DeploymentCommand
+  ): Future[Either[RunDeploymentError, Future[Option[ExternalDeploymentId]]]] = {
+    command match {
+      case command: RunDeploymentCommand =>
+        runDeployment(command)
+    }
+  }
+
+  private def runDeployment(
+      command: RunDeploymentCommand
+  ): Future[Either[RunDeploymentError, Future[Option[ExternalDeploymentId]]]] = {
+    dbioRunner.runInTransactionE(
+      (for {
+        scenarioMetadata <- getScenarioMetadata(command)
+        _ <- EitherT.fromEither(
```

Comment:
I can be simplified:
```scala
_ <- EitherT.cond(command.user.can(scenarioMetadata.processCategory, Permission.Deploy), (), NoPermissionError)
```

BTW, we could also create a `checkPermission` private method to make it more readable

Reply:
Simplified. This private method would take: user, category, expectedPermission and it is one extra jump to understand what's going on. I think that it's an overhead ;-)

Reply:
IMO it's more readable:

```scala
  private def runDeployment(
      command: RunDeploymentCommand
  ): Future[Either[RunDeploymentError, Unit]] =
    dbioRunner.run(
      (for {
        scenarioMetadata <- getScenarioMetadata(command)
        _ <- checkPermission(user = command.user, category = scenarioMetadata.processCategory, permission = Deploy)
        _ <- saveDeployment(command, scenarioMetadata)
        runResult <- invokeLegacyRunDeploymentLogic(command, scenarioMetadata)
      } yield runResult).value
    )

  private def checkPermission(user: LoggedUser, category: String, permission: Permission) = {
    EitherT.cond[DB](user.can(category, Deploy), (), NoPermissionError)
  }
```

I like how the EitherT simplifies the for-comprehension, but it's sometimes too verbose, and encapsulating it in the private method in most cases helps to get rid of the noise which is not important for the logic.

But I leave it the decision to you.

Reply:
After wrapping the line, it looks pretty much the same ;-) But I changed it if it looks better for you.